src/martok/Martok.ts

```diff
@@ -51,11 +65,59 @@ export class Martok {
     return this.storage.getStore()!.typeReplacer;
   }
 
-  private readonly storage = new AsyncLocalStorage<MartokState>();
-  private readonly imports = new ImportGenerator(this);
-  private readonly formatter = new MartokFormatter(this.config);
+  private readonly storage;
+  private readonly imports;
+  private readonly flattener;
+  private readonly formatter;
+
+  public constructor(public readonly config: MartokConfig) {
+    // Get all file raw text
+    for (const file of this.config.files) {
+      try {
+        const result = fs.readFileSync(file, "utf-8");
+        this.fsMap.set(file, result);
+      } catch (e) {
+        console.error(`Failed to read file ${file}: `, e);
+      }
+    }
 
-  public constructor(public readonly config: MartokConfig) {}
+    // Create initial program
+    const { program, env } = this.compileFiles(this.fsMap);
+    this.env = env;
+    this.program = program;
+
+    this.flattener = new Flattener(this);
+
+    if (this.config.options?.experimentalTypeFlattening) {
+      console.log("Flattening types...");
+      // Flatten all the files, recompile them together, then swap out our program
+      const { program, fs, env } = this.flattener.flattenFileSystem();
+      this.fsMap = fs;
+      this.env = env;
+      this.program = program;
+    }
+
+    this.declarations = new DeclarationGenerator(this);
```

Comment:
This switch away from in-line member initialization, is fine, but it does represent a marked shift away from the prevailing style of the project. It's worth a discussion if we're going to make the change.

Reply:
For these changes to work, it's definitely important to know when members get initialized, i've not really been a fan of in-line initialization because the order is a bit more ambiguous. If you have strong feelings against it, totally open to talk about it

Reply:
Can you point to specific things that would have broken if we stayed with inline init?

Reply:
Lines 100-104 in `martok.ts` weren't working because they were being initialized before flattening occurred I *think*. There was a reason I moved them and i'm pretty sure that's it. I'll move them back and if everything works then I we can leave things the way they were